### PR TITLE
Add phantom instrumentation

### PR DESCRIPTION
**Pull request by michaelpigg**
Adding support for Scala Phantom library https://github.com/outworkers/phantom

Strategy is to add a completion listener to the Future returned from the `future` method.

TODO:
- [ ] Fix unexpected active scope (need for shutdownScopes in test) https://github.com/DataDog/dd-trace-java/pull/1515/files#diff-52752f1e85dda8da098822ca4c7aeb47R127
- [ ] More tests on common operations
- [ ] Add latestDepTest
- [ ] Add muzzle
- [ ] code formatting

### CODE REVIEW

File: dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      trace(0, 3) {
+        basicSpan(it, 0, "parent")
+        phantomSpan(it, 1, cql, null, it.span(0), null)
+        cassandraSpan(it, 2, cql, null, false, it.span(1))
+      }
+    }
+
+    where:
+    id                       | cmd             | cql                                                                                                | ec
+    UUID.randomUUID()        | insertBook      | "UPDATE books.books SET title = 'Programming in Scala', author = 'Odersky' WHERE id = " + id + ";" | scala.concurrent.ExecutionContext$.MODULE$.global()
```
Comment (devinsba) [inline dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy:74]: Is the cassandra instrumentation adding in the query parameters here, or is this the phantom instrumentation? If it's phantom we we would not want to do that by default. We don't want users potentially sending secrets to us without realizing it

File: dd-java-agent/instrumentation/phantom/phantom.gradle [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  //compile project(':dd-java-agent:instrumentation:hibernate')
+  testCompile project(':dd-java-agent:instrumentation:datastax-cassandra-3')
+  testCompile project(':dd-java-agent:instrumentation:phantom')
+  testCompile group: 'com.outworkers', name: 'phantom-dsl_2.11', version: '2.59.0'
+  testCompile group: 'com.datastax.cassandra', name: 'cassandra-driver-core', version: '3.0.0'
+
+  testCompile deps.scala
+  testCompile group: 'org.scala-lang', name: 'scala-reflect', version: '2.11.12'
+  testCompile group: 'org.cassandraunit', name: 'cassandra-unit', version: '3.11.2.0'
+}
```
Comment (devinsba) [inline dd-java-agent/instrumentation/phantom/phantom.gradle:30]: Generally we like to have a latestDepTest task that would test the latest version of phantom that this instrumentation supports

File: dd-java-agent/instrumentation/phantom/phantom.gradle [hunk 3]
```diff
@@ -0,0 +1,31 @@
+// Set properties before any plugins get loaded
+ext {
+  minJavaVersionForTests = JavaVersion.VERSION_1_8
+}
+
+apply from: "${rootDir}/gradle/java.gradle"
+apply from: "${rootDir}/gradle/test-with-scala.gradle"
+
+apply plugin: 'org.unbroken-dome.test-sets'
+
```
Comment (devinsba) [inline dd-java-agent/instrumentation/phantom/phantom.gradle:10]: Ah, I missed this. We also need a muzzle check. This gives us a test that the instrumentation activates when we expect it to, and optionally doesn't activate when we expect it not to

File: dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy (hunk 1 above)
Comment (michaelpigg) [reply dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy:74]: I was wondering about this. Is there something in the cassandra instrumentation that avoids this issue?

**Comment by michaelpigg** [comment]
@devinsba Can you help me understand why I get the error
 ```java.lang.AssertionError: Span is active before test has started```
in the test? It seems that the span from the previous test is being found, but I'm not clear why it would still be active.

https://app.circleci.com/pipelines/github/DataDog/dd-trace-java/1216/workflows/e6917aa6-9bf9-4ec1-923f-ae320e29a966/jobs/64994/tests

File: dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+import static datadog.trace.agent.test.utils.TraceUtils.basicSpan
+import static datadog.trace.agent.test.utils.TraceUtils.runUnderTrace
+
+class PhantomInstrumentationTest extends AgentTestRunner {
+
+  @Shared
+  Cluster cluster
+  @Shared
+  int port = 9142
```
Comment (randomanderson) [inline dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy:20]:
```suggestion
  int port = PortUtils.randomOpenPort();
```

**Comment by tylerbenson** [comment]
@michaelpigg Do you need help finishing this up?

File: dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy (hunk 4 above)
Comment (michaelpigg) [reply dd-java-agent/instrumentation/phantom/src/test/groovy/PhantomInstrumentationTest.groovy]: Updated to use random port

**Comment by michaelpigg** [comment]
> @michaelpigg Do you need help finishing this up?

@tylerbenson I think my main issue is https://github.com/DataDog/dd-trace-java/pull/1515#issuecomment-637064505. I can't figure out why the phantom span doesn't seem to finish cleanly. The scope manager thinks it's still active when the test tries to close the "parent" span.

**Comment by tylerbenson** [comment]
@michaelpigg thanks for the update.  @devinsba will take a look at this next week.

**Comment by devinsba** [comment]
@michaelpigg checking this morning

**Comment by tylerbenson** [comment]
@michaelpigg can we close this in favor of #1665 since that seems to be the underlying problem?

**Comment by michaelpigg** [comment]
@tylerbenson Closing for now. May want to bring it back at some point.

**Pull request closed without merging**